### PR TITLE
Query and update animals' table

### PR DESCRIPTION
**Pull request by diarisdiakite**
In this PR, I:
✔️ Added a column species of type string to the animals' table and modified the schema.sql file.
✔️ Inserted data in the table
✔️ Added these insertions to my data.sql file. (see screenshot)

image

Inside a transaction I :
✔️ Updated the animals table by setting the species column to unspecified. Verified that change was made. Then rolled back the change and verified that the species columns went back to the state before the transaction.

image

Inside a transaction, I :
✔️ Updated the animals table by setting the species column to digimon for all animals that have a name ending in mon.
✔️ Updated the animals table by setting the species column to pokemon for all animals that don't have species already set.
✔️ Verified that changes were made.
✔️ Commited the transaction.
✔️ Verified that changes persist after commit.
✔️ Took a screenshot of the results of the actions.

image

Inside a transaction, I :
✔️ Deleted all records in the animals' table, then rolled back the transaction.
✔️Verified if all records in the animals' table still exist after the Rollback.
✔️ Took a screenshot of the results of the actions.

image

✔️ Deleted all animals born after Jan 1st, 2022.
✔️ Created a savepoint for the transaction.
✔️ Updated all animals' weight to be their weight multiplied by -1.
✔️ Rolled back to the savepoint
✔️ Updated all animals' weights that are negative to be their weight multiplied by -1.
✔️ Commited transaction
✔️ Took a screenshot of the results of the actions.
✔️ Added all these transactions to the queries.sql file.

image

image

image

### CODE REVIEW

**Comment by diarisdiakite** [comment]
In this milestone, I:
Inside a transaction I :
✔️ Updated the animals table by setting the species column to unspecified. Verified that change was made. Then rolled back the change and verified that the species columns went back to the state before the transaction.

image

Inside a transaction, I :
✔️ Updated the animals table by setting the species column to digimon for all animals that have a name ending in mon.
✔️ Updated the animals table by setting the species column to pokemon for all animals that don't have species already set.
✔️ Verified that changes were made.
✔️ Commited the transaction.
✔️ Verified that changes persist after commit.
✔️ Took a screenshot of the results of the actions.

image

Inside a transaction, I :
✔️ Deleted all records in the animals' table, then rolled back the transaction.
✔️Verified if all records in the animals' table still exist after the Rollback.
✔️ Took a screenshot of the results of the actions.

image


✔️ Deleted all animals born after Jan 1st, 2022.
✔️ Created a savepoint for the transaction.
✔️ Updated all animals' weight to be their weight multiplied by -1.
✔️ Rolled back to the savepoint
✔️ Updated all animals' weights that are negative to be their weight multiplied by -1.
✔️ Commited transaction
✔️ Took a screenshot of the results of the actions.
✔️ Added all these transactions to the queries.sql file.

image


image


image

**Comment by diarisdiakite** [comment]
In this milestone, I:
✔️ Wrote queries to answer the following questions:
✔️ How many animals are there?


✔️ How many animals have never tried to escape?


✔️ What is the average weight of animals?


✔️ Who escapes the most, neutered or not neutered animals?


✔️ What is the minimum and maximum weight of each type of animal?


✔️ What is the average number of escape attempts per animal type of those born between 1990 and 2000?


✔️ Took a screenshot of the results of the queries.
✔️ Added these queries to my queries.sql file.
✔️ Added screenshots with the results of the queries and with the results of the play with transactions to the Pull Request description.

**Comment by diarisdiakite** [comment]
In this milestone, I wrote queries to answer the following questions:
✔️ How many animals are there?
✔️ How many animals have never tried to escape?
✔️ What is the average weight of animals?
✔️ Who escapes the most, neutered or not neutered animals?
image


✔️ What is the minimum and maximum weight of each type of animal?
✔️ What is the average number of escape attempts per animal type of those born between 1990 and 2000?
image


✔️ Took screenshots of the results of the different queries.
✔️ Added these queries to my queries.sql file.
✔️ Added screenshots with the results of the queries and with the results of the play with transactions to the Pull Request description.

File: data.sql [hunk 1]
```diff
@@ -10,3 +10,18 @@ VALUES
...
+
+/* Insertion 2 */
+INSERT INTO animals (name, date_of_birth, escape_attempts, neutered, weight_kg, species)
+VALUES 
+  ('Charmander', '2020-02-08', 0, false, 11, null),
+  ('Plantmon', '2021-11-15', 2, true, 5.7, null),
+  ('Squirtle', '1993-04-02', 3, false, 12.13, null),
+  ('Angemon', '2005-06-12', 1, true, 45, null),
+  ('Boarmon', '2005-06-07', 7, true, 20.4, null),
+  ('Blossom', '1998-10-13', 3, true, 17, null),
```
Comment (Davidosky007) [inline data.sql]: * Kindly adjust the weights as some of the required weights are negative numbers kindly check the screenshot below 👇 
image

File: data.sql [hunk 2]
```diff
@@ -10,3 +10,18 @@ VALUES
   ('Gabumon', '2018-11-15', 2, true, 8),
   ('Pikachu', '2021-01-07', 1, false, 15.04),
   ('Devimon', '2017-05-12', 5, true, 11);
+
+
+/* Add species column to the animals table */
+ALTER TABLE ANIMALS ADD SPECIES VARCHAR(50);
```
Comment (Davidosky007) [inline data.sql]: * Kindly note that Everything that creates or changes the table structure should be saved in the schema.sql file. [check requirements](https://github.com/microverseinc/curriculum-databases/blob/main/db-structure/vet_clinic_query_update_tables.md#:~:text=Add%20a%20column%20species%20of%20type%20string%20to%20your%20animals%20table.%20Modify%20your%20schema.sql%20file.)

File: queries.sql [hunk 3]
```diff
@@ -27,3 +27,86 @@ SELECT * FROM ANIMALS WHERE NAME != 'Gabumon';
...
+  transactions 
+
+*/
+
+/* Update the animals table by setting the species column to unspecified. Verify that change was made. Then rolled back the change and verified that the species columns went back to the state before the transaction. */
+
+BEGIN;
+UPDATE ANIMALS SET SPECIES = 'unspecified';
+SELECT * FROM ANIMALS;
+ROLLBACK;
```
Comment (Davidosky007) [inline queries.sql:50]: * As stated in the [requirement](https://github.com/microverseinc/curriculum-databases/blob/main/db-structure/vet_clinic_query_update_tables.md#:~:text=verify%20that%20the%20species%20columns%20went%20back%20to%20the%20state%20before%20the%20transaction.), Kindly verify that the species columns went back to the state before the transaction after rolling back the change. HINT 💭 running these queries  again  (`SELECT species from animals;`) verifies that the species column went back to original state

File: queries.sql [hunk 4]
```diff
@@ -27,3 +27,86 @@ SELECT * FROM ANIMALS WHERE NAME != 'Gabumon';
...
+Update the animals table by setting the species column to digimon for all animals that have a name ending in mon.
+Update the animals table by setting the species column to pokemon for all animals that don't have species already set.
+Verify that changes were made.
+Commit the transaction. */
+
+BEGIN;
+UPDATE ANIMALS SET SPECIES = 'digimon' WHERE NAME LIKE '%mon';
+UPDATE ANIMALS SET SPECIES = 'pokemon' WHERE SPECIES IS NULL OR SPECIES = '';
+SELECT * FROM ANIMALS;
+COMMIT;
```
Comment (Davidosky007) [inline queries.sql:63]: * Also here You are [required](https://github.com/microverseinc/curriculum-databases/blob/main/db-structure/vet_clinic_query_update_tables.md#:~:text=Verify%20that%20changes%20persist%20after%20commit.) to Verify that changes persist after commit, Kindly check the review comments above for Hints ⬆️

**Review by Davidosky007 — CHANGES_REQUESTED** [review]
Hi @diarisdiakite 👋, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
* Well-written PR 👍 
* Database transactions used properly 👍 
* Data modified properly using SQL 👍 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @davidsoky007 in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: data.sql (hunk 1 above)
Comment (diarisdiakite) [reply data.sql]: Thank you @Davidosky007, changes have been made.

File: data.sql (hunk 2 above)
Comment (diarisdiakite) [reply data.sql]: Done.

File: queries.sql (hunk 3 above)
Comment (diarisdiakite) [reply queries.sql:50]: I had added the checking for all the transactions in my screenshots queries with a "select *" query. Adding them to the queries.sql as well then.

File: data.sql (hunk 1 above)
Comment (diarisdiakite) [reply data.sql]: image

Comment (diarisdiakite) [reply data.sql]: and in the queries file as well
image

**Comment by diarisdiakite** [comment]
In this milestone, I made changes to adress the code reviewer's comments. Here are screenshots of the affected queries. 
image

✔️ Updated the animals table by setting the species column to unspecified. Verified that change was made. Then rolled back the change and verified that the species columns went back to the state before the transaction.
image

✔️ Update the species Update by setting the column to digimon for all animals that have a name ending in mon.
and to pokemon for all animals that don't have species already set. Verified that changes were made and Committed the transaction.
image

Inside a transaction:
✔️ Deleted all animals born after Jan 1st, 2022. (was already done)
✔️ Created a savepoint for the transaction.
✔️ Updated all animals' weight to be their weight multiplied by -1.
✔️ Rolled back to the savepoint
✔️ Updated all animals' weights that are negative to be their weight multiplied by -1.
✔️ Commited transaction
✔️ Took a screenshot of the results of your actions. 
image

**Comment by diarisdiakite** [comment]
In this milestone, I made changes to adress the code reviewer's comments. Here are screenshots of the affected queries. 
image

✔️ Updated the animals table by setting the species column to unspecified. Verified that change was made. Then rolled back the change and verified that the species columns went back to the state before the transaction.
image

✔️ Update the species Update by setting the column to digimon for all animals that have a name ending in mon.
and to pokemon for all animals that don't have species already set. Verified that changes were made and Committed the transaction.
image

Inside a transaction:
✔️ Deleted all animals born after Jan 1st, 2022. (was already done)
✔️ Created a savepoint for the transaction.
✔️ Updated all animals' weight to be their weight multiplied by -1.
✔️ Rolled back to the savepoint
✔️ Updated all animals' weights that are negative to be their weight multiplied by -1.
✔️ Commited transaction
✔️ Took screenshots of the results of your actions. 
image

image

**Review by Davidosky007 — APPROVED** [review]
Hi @diarisdiakite 👋, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
* Well-written PR 👍
* Database transactions used properly 👍
* Data modified properly using SQL 👍
## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**